crates/ruff_linter/src/rules/perflint/rules/manual_list_comprehension.rs

```diff
@@ -188,10 +203,159 @@ pub(crate) fn manual_list_comprehension(checker: &mut Checker, for_stmt: &ast::S
         return;
     }
 
-    checker.diagnostics.push(Diagnostic::new(
+    let binding_stmt = binding
+        .statement(checker.semantic())
+        .and_then(|stmt| stmt.as_assign_stmt());
+
+    // If the variable is an empty list literal, then we might be able to replace it with a full list comprehension
+    // otherwise, it has to be replaced with a `list.extend`
+    let binding_is_empty_list =
+        binding_stmt.is_some_and(|binding_stmt| match binding_stmt.value.as_list_expr() {
+            Some(list_expr) => list_expr.elts.is_empty(),
+            None => false,
+        });
+
+    let comprehension_type = if binding_is_empty_list {
+        ComprehensionType::ListComprehension
+    } else {
+        ComprehensionType::Extend
+    };
+
+    // If the for loop does not have the same parent element as the binding, then it cannot always be
+    // deleted and replaced with a list comprehension. This does not apply when using an extend.
+    let assignment_in_same_statement = {
+        let for_loop_parent = checker.semantic().current_statement_parent_id();
+
+        binding.source.is_some_and(|binding_source| {
+            let binding_parent = checker.semantic().parent_statement_id(binding_source);
+            for_loop_parent == binding_parent
+        })
+    };
+
+    // If the binding is not a single name expression, it could be replaced with a list comprehension,
+    // but not necessarily, so this needs to be manually fixed. This does not apply when using an extend.
+    let binding_has_one_target = {
+        let only_target = binding_stmt.is_some_and(|binding_stmt| binding_stmt.targets.len() == 1);
+        let is_name =
+            binding_stmt.is_some_and(|binding_stmt| binding_stmt.targets[0].is_name_expr());
+        only_target && is_name
+    };
+
+    // A list extend works in every context, while a list comprehension only works when all the criteria are true
+    let comprehension_type =
+        Some(comprehension_type).filter(|comprehension_type| match comprehension_type {
+            ComprehensionType::ListComprehension => {
+                binding_stmt.is_some() && assignment_in_same_statement && binding_has_one_target
+            }
+            ComprehensionType::Extend => true,
+        });
+
+    let mut diagnostic = Diagnostic::new(
         ManualListComprehension {
             is_async: for_stmt.is_async,
+            comprehension_type,
         },
         *range,
-    ));
+    );
+
+    diagnostic.try_set_optional_fix(|| match comprehension_type {
```

Comment:
Let's gate the new fix behind preview `if checker.settings.preview.is_enabled() { ... }`